### PR TITLE
MCP39F511N metrics provider

### PR DESCRIPTION
**Pull request by ArneTR**
@ribalba Can you give the C code a quick read.

As said in the file: I am including a library of unknown original origin.

    This file is mostly copied from https://github.com/osmhpi/pinpoint/blob/master/src/data_sources/mcp_com.c
    Credits to Sven Köhler and the OSM group from the HPI

    In case the file is not original work: Possible prior origins of the file are unknown. However it is an implementation
    of the protocol defined here:
    - https://www.microchip.com/en-us/development-tool/ADM00706
    - https://ww1.microchip.com/downloads/aemDocuments/documents/OTH/ProductDocuments/DataSheets/20005473B.pdf

### CODE REVIEW

File: metric_providers/psu/energy/ac/mcp/machine/source.c [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    unsigned char command_packet[80];
+    int rdlen;
+    uint8_t len;
+    uint8_t i;
+    uint8_t checksum = 0;
+    uint8_t datap = 0;
+
+    command_packet[0] = 0xa5;
+    command_packet[1] = cmd_length + 3;
+    memcpy(command_packet + 2, cmd, cmd_length);
```
Comment (ribalba) [inline metric_providers/psu/energy/ac/mcp/machine/source.c:98]: `cmd` and `cmd_length` are not checked if they fit into cmd

File: metric_providers/psu/energy/ac/mcp/machine/source.c [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    for (i = 0; i < cmd_length + 2; i++) {
+        checksum += command_packet[i];
+    }
+    command_packet[i] = checksum;
```
Comment (ribalba) [inline metric_providers/psu/energy/ac/mcp/machine/source.c:102]: why if `cmd_length` is > 80

File: metric_providers/psu/energy/ac/mcp/machine/source.c [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    tcflush(fd, TCIOFLUSH);
+    len = write(fd, command_packet, cmd_length + 3);
```
Comment (ribalba) [inline metric_providers/psu/energy/ac/mcp/machine/source.c:104]: Why is fd passed as a parameter if it is defined globally?

File: metric_providers/psu/energy/ac/mcp/machine/source.c [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    { 0x41, 0x00, 0x40, 0x52 };
+const unsigned char f501_set_apparent_power_divisor[] =
+    { 0x41, 0x00, 0x40, 0x57, 0x00, 0x03 };
+const unsigned char f501_set_accumulation_interval[] =
+    { 0x41, 0x00, 0x5A, 0x57, 0x00, 0x00 };
+const unsigned char f501_read_range[] = { 0x41, 0x00, 0x48, 0x44 };
+
+const unsigned char f511_read_active_power[] = { 0x41, 0x0, 0x16, 0x4E, 8 };
+const unsigned char f511_read_active_power1[] = { 0x41, 0x0, 0x16, 0x4E, 4 };
+const unsigned char f511_read_active_power2[] = { 0x41, 0x0, 0x1a, 0x4E, 4 };
```
Comment (ribalba) [inline metric_providers/psu/energy/ac/mcp/machine/source.c:35]: What is the difference between these read power chars?

**Review by ribalba — CHANGES_REQUESTED** [review]
Just an initial quick read. Is this something we want to maintain for the future? The I would probably refactor it. I can do this :)

**Comment by ArneTR** [comment]
Updated. @ribalba Can you give it another look?

I clarified in the code what the original author is doing, although some implementations are not documented and cannot be checked completely

**Comment by ArneTR** [comment]
I will merge it in for now, please put your comments in here if you have any

**Pull request merged**